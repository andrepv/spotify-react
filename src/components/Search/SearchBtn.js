import React, { Component } from "react";
import PropTypes from "prop-types";
import { Search as Icon } from "react-feather";
import { connectSearchResults } from "containers/SearchResultsContainer";

export class SearchBtn extends Component {
  render() {
    const {toggleSearch, renderContent, className} = this.props;
    if (renderContent) {
      return renderContent(toggleSearch, <Icon />);
    }
    return (
      <div onClick={toggleSearch}>
        <Icon className={className} />
      </div>
    );
  }
}

SearchBtn.propTypes = {
  toggleSearch: PropTypes.func.isRequired,
  renderContent: PropTypes.func,
  className: PropTypes.string,
};

export default connectSearchResults(SearchBtn);