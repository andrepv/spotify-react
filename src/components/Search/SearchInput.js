import React, { Component } from "react";
import PropTypes from "prop-types";
import { X } from "react-feather";
import { ReactComponent as Loader } from "images/loader.svg";

export default class SearchInput extends Component {
  shouldComponentUpdate(nextProps) {
    return (
      nextProps.pending !== this.props.pending ||
      nextProps.value !== this.props.value
    );
  }

  render() {
    const {pending, value, updateSearchValue, close} = this.props;
    return (
      <div className="search__input-container flex-center">
        <div className="search__input-wrapper">
          <p>What are you looking for?</p>
          <div className="search__input">
            <input
              onChange={updateSearchValue}
              type='text'
              placeholder='Start Typing...'
              className="search__input-field"
              value={value}
            />
            {pending &&
              <span className="search__loader"><Loader /></span>
            }
            <span
              className="search__close"
              onClick={close}
            >
              <X/>
            </span>
          </div>
        </div>
      </div>
    );
  }
}

SearchInput.propTypes = {
  close: PropTypes.func.isRequired,
  updateSearchValue: PropTypes.func.isRequired,
  pending: PropTypes.bool,
  value: PropTypes.string,
};