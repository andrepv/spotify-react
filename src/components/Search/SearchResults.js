import React, { Component } from "react";
import PropTypes from "prop-types";

import Tracks from "components/Common/Tracks";
import Artist from "components/Common/Artist";
import Block from "components/Common/Block";

export default class SearchResults extends Component {
  componentDidMount() {
    document.body.classList.add("disable-scroll");
  }

   componentWillUnmount() {
    document.body.classList.remove("disable-scroll");
  }

  renderNotFoundMsg = () => {
    return (
      <div className="search__not-found">
        <h2>{`No results found for "${this.props.value}"`}</h2>
        <p>Please make sure your words are spelled correctly
        or use less or different keywords</p>
      </div>
    );
  }

  renderTabs() {
    const {type, switchTab} = this.props;
    const options = ["track", "album", "playlist", "artist"];
    return options.map((option, index) => {
      return (
        <li
          key={index}
          onClick={() => switchTab(option)}
          className={`search__tab ${option === type
            ? "search__tab_active": ""
          }`}
        >
          {`${option}s`}
        </li>
      );
    });
  }

  renderTracks() {
    const {items} = this.props;
    return (
      <div className="tracks tracks_col-3">
        <Tracks
          columns={3}
          trackList={items}
          source={{name: "SearchResults"}}
        />
      </div>
    );
  }

  renderArtists() {
    const {items, close} = this.props;
    return (
      <div className="artists artists_cols-6">
        {items.map(item => {
          return (
            <Artist
              key={item.id}
              id={item.id}
              image={item.image}
              name={item.name}
              handler={close}
            />
          );
        })}
      </div>
    );
  }

  renderBlocks() {
    const {items, type, close} = this.props;
    return (
      <div className="blocks-container blocks_cols-5">
        {items.map(item => {
          return (
            <Block
              key={item.id}
              id={item.id}
              meta={item.meta}
              type={type}
              name={item.name}
              image={item.image}
              handler={close}
            />
          );
        })}
      </div>
    );
  }

  render() {
    const {items, pending, type} = this.props;
    const content = {
      track: () => {
        return this.renderTracks();
      },
      artist: () => {
        return this.renderArtists();
      },
      album: () => {
        return this.renderBlocks();
      },
      playlist: () => {
        return this.renderBlocks();
      },
    };
    return (
      <div className="search__results custom-scrollbar">
        <ul className="search__tabs">
          {this.renderTabs()}
        </ul>
        {pending
          ? null
          : !items.length
            ? this.renderNotFoundMsg()
            : content[type]()
        }
      </div>
    );
  }
}

SearchResults.propTypes = {
  value: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  switchTab: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  close: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired,
};