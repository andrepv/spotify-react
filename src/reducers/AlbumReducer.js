import {
  ALBUM,
  NEW_RELEASES,
  ARTIST_ALBUMS,
  ARTIST_SINGLES,
  ADD_TO_MY_SAVED_ALBUMS,
  REMOVE_FROM_MY_SAVED_ALBUMS,
  ADD_TO_FAVORITE_TRACKS,
  REMOVE_FROM_FAVORITE_TRACKS,
} from "constants/ActionConstants";
import toggleLike from "utils/toggleLike";

const initialNewReleasesState = {
  pending: false,
  loadMorePending: false,
  items: [],
  total: 0,
  error: false,
};

const initialAlbumState = {
  pending: false,
  authors: [],
  id: "",
  image: "",
  name: "",
  date: "",
  totalTracks: 0,
  isSaved: false,
  items: [],
  isAvailableForPreview: false,
  error: false,
};

const initialArtistAlbumsState = {
  pending: false,
  loadMorePending: false,
  total: 0,
  items: [],
  error: false,
};

const initialArtistSinglesState = {
  pending: false,
  loadMorePending: false,
  items: [],
  total: 0,
  error: false,
};

export function newReleases(state = initialNewReleasesState, action) {
  switch (action.type) {
    case NEW_RELEASES.PENDING:
      return {
        ...state,
        pending: true,
      };
    case NEW_RELEASES.SUCCESS:
      return {
        ...state,
        pending: false,
        items: action.payload.albums,
        total: action.payload.total,
      };
    case NEW_RELEASES.ERROR:
      return {
        ...state,
        pending: false,
        loadMorePending: false,
        error: true,
      };
    case NEW_RELEASES.LOAD_MORE_PENDING:
      return {
        ...state,
        loadMorePending: true,
      };
    case NEW_RELEASES.LOAD_MORE_SUCCESS:
      return {
        ...state,
        loadMorePending: false,
        items: state.items.concat(action.payload),
      };
    case NEW_RELEASES.LOAD_MORE_ERROR:
      return {
        ...state,
        loadMorePending: false,
      };
    default:
      return state;
  }
}

export function album(state = initialAlbumState, action) {
  switch (action.type) {
    case ALBUM.PENDING:
      return {
        ...state,
        pending: true,
      };
    case ALBUM.SUCCESS:
      return {
        ...state,
        pending: false,
        authors: action.payload.artists,
        id: action.payload.id,
        image: action.payload.images[0].url,
        name: action.payload.name,
        date: action.payload.release_date.split("-")[0],
        totalTracks: action.payload.totalTracks,
        items: action.payload.tracks,
        isAvailableForPreview: action.payload.isAvailableForPreview,
        isSaved: action.payload.isSaved,
      };
    case ADD_TO_MY_SAVED_ALBUMS:
      return {
        ...state,
        isSaved: true,
      };
    case REMOVE_FROM_MY_SAVED_ALBUMS:
      return {
        ...state,
        isSaved: false,
      };
    case ADD_TO_FAVORITE_TRACKS:
      return {
        ...state,
        items: toggleLike(state.items, action.payload.id),
      };
    case REMOVE_FROM_FAVORITE_TRACKS:
      return {
        ...state,
        items: toggleLike(state.items, action.payload.id),
      };
    case ALBUM.ERROR:
      return {
        ...state,
        pending: false,
        error: true,
      };
    default:
      return state;
  }
}

export function artistAlbums(state = initialArtistAlbumsState, action) {
  switch (action.type) {
    case ARTIST_ALBUMS.PENDING:
      return {
        ...state,
        pending: true,
      };
    case ARTIST_ALBUMS.SUCCESS:
      return {
        ...state,
        pending: false,
        items: action.payload.items,
        total: action.payload.total,
      };
    case ARTIST_ALBUMS.ERROR:
      return {
        ...state,
        pending: false,
        error: true,
      };
    case ARTIST_ALBUMS.LOAD_MORE_PENDING:
      return {
        ...state,
        loadMorePending: true,
      };
    case ARTIST_ALBUMS.LOAD_MORE_SUCCESS:
      return {
        ...state,
        loadMorePending: false,
        items: state.items.concat(action.payload),
      };
    case ARTIST_ALBUMS.LOAD_MORE_ERROR:
      return {
        ...state,
        loadMorePending: false,
      };
    default:
      return state;
  }
}

export function artistSingles(state = initialArtistSinglesState, action) {
  switch (action.type) {
    case ARTIST_SINGLES.PENDING:
      return {
        ...state,
        pending: true,
      };
    case ARTIST_SINGLES.SUCCESS:
      return {
        ...state,
        pending: false,
        items: action.payload.items,
        total: action.payload.total,
      };
    case ARTIST_SINGLES.ERROR:
      return {
        ...state,
        pending: false,
        error: true,
      };
    case ARTIST_SINGLES.LOAD_MORE_PENDING:
      return {
        ...state,
        loadMorePending: true,
      };
    case ARTIST_SINGLES.LOAD_MORE_SUCCESS:
      return {
        ...state,
        loadMorePending: false,
        items: state.items.concat(action.payload),
      };
    case ARTIST_SINGLES.LOAD_MORE_ERROR:
      return {
        ...state,
        loadMorePending: false,
      };
    default:
      return state;
  }
}