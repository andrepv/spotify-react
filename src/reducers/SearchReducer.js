import {
  FILTER_BY_TYPE,
  SEARCH_RESULTS,
  RESET_SEARCH_RESULTS,
  TOGGLE_SEARCH,
  ADD_TO_FAVORITE_TRACKS,
  REMOVE_FROM_FAVORITE_TRACKS,
} from "constants/ActionConstants";
import toggleLike from "utils/toggleLike";

export const initialSearchResultsState = {
  pending: false,
  items: [],
  type: "track",
  isOpen: false,
  error: false,
};

export default function searchResults(
  state = initialSearchResultsState, action
) {
  switch ( action.type ) {
    case FILTER_BY_TYPE:
      return {
        ...state,
        type: action.payload,
      };
    case SEARCH_RESULTS.PENDING:
      return {
        ...state,
        pending: true,
      };
    case SEARCH_RESULTS.SUCCESS:
      return {
        ...state,
        pending: false,
        items: action.payload,
      };
    case SEARCH_RESULTS.ERROR:
      return {
        ...state,
        pending: false,
        error: true,
      };
    case RESET_SEARCH_RESULTS:
      return {
        pending: false,
        items: [],
        type: "track",
        isOpen: state.isOpen,
        error: false,
      };
    case TOGGLE_SEARCH:
      return {
        ...state,
        isOpen: !state.isOpen,
      };
    case ADD_TO_FAVORITE_TRACKS:
      return {
        ...state,
        items: toggleLike(state.items, action.payload.id),
      };
    case REMOVE_FROM_FAVORITE_TRACKS:
      return {
        ...state,
        items: toggleLike(state.items, action.payload.id),
      };
    default:
      return state;
  }
}