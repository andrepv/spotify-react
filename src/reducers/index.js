import { combineReducers } from "redux";

import {
  categoryPlaylists,
  featuredPlaylists,
  userPlaylists,
  listUserPlaylists,
  playlist,
} from "./PlaylistReducer";

import {
  newReleases,
  album,
  artistAlbums,
  artistSingles,
} from "./AlbumReducer";

import {
  followedArtists,
  artist,
  relatedArtists,
} from "./ArtistReducer";

import categories from "./CategoriesReducer";
import charts from "./ChartsReducer";
import searchResults from "./SearchReducer";

import {
  player,
  progressBar,
} from "./PlayerReducer";

import {
  artistTopTracks,
  topTracks,
  savedTracks,
} from "./TrackReducer";

export const rootReducer = combineReducers({
  userPlaylists,
  featuredPlaylists,
  categoryPlaylists,
  playlist,
  listUserPlaylists,

  newReleases,
  album,
  artistAlbums,
  artistSingles,

  relatedArtists,
  followedArtists,
  artist,

  topTracks,
  savedTracks,
  artistTopTracks,

  categories,
  player,
  progressBar,
  charts,
  searchResults,
});