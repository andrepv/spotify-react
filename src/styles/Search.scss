@import
"variables",
"modularscale";

.search {
  &__input {
    width: 100%;
    position: relative;
    display: flex;
    align-items: center;
    justify-content: space-between;
    &-container {
      position: fixed;
      z-index: 999;
      background: $bg-primary;
      top: 0;
      left: 0;
      width: 100%;
      height: 30vh;
      padding: 0 2em;
      box-shadow: $shadows;
      &-enter {
        transform: translateY(-200%);
      }
      &-enter-active {
        transform: translateY(0);
        transition: all 600ms ease-in-out;
      }
      &-exit {
        transform: translateY(0);
      }
      &-exit-active {
        transform: translateY(-200%);
        transition: all 600ms ease-in-out;
      }
    }
    &-wrapper {
      width: 100%;
      height: 95px;
      display: flex;
      flex-wrap: wrap;
      justify-content: space-between;
      align-items: center;
      color: $text-secondary;
    }
    &-field {
      font-size: ms(16);
      border: 0;
      width: 90%;
    }
  }
  &__close {
    cursor: pointer;
    svg {
      width: 40px;
      height: 40px;
    }
  }
  &__results {
    position: fixed;
    z-index: 999;
    background: $bg-secondary;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 70vh;
    padding: 40px;
    overflow-y: scroll;
    overflow-x: hidden;
    @media (max-width: 450px) {
      padding: 40px 20px;
    }
    &-enter {
      transform: translateY(200%);
    }
    &-enter-active {
      transform: translateY(0);
      transition: all 600ms ease-in-out;
    }
    &-exit {
      transform: translateY(0);
    }
    &-exit-active {
      transform: translateY(200%);
      transition: all 600ms ease-in-out;
    }
  }
  &__tabs {
    margin-bottom: 40px;
  }
  &__tab {
    display: inline-block;
    margin-right: 20px;
    font-size: ms(1);
    cursor: pointer;
    text-transform: capitalize;
    &_active {
      border-bottom: 2px solid;
    }
  }
}