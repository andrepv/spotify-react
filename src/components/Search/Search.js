import React, { Component } from "react";
import PropTypes from "prop-types";
import { CSSTransition, TransitionGroup } from "react-transition-group";

import { connectSearchResults } from "containers/SearchResultsContainer";
import SearchResults from "./SearchResults";
import SearchInput from "./SearchInput";
import "styles/Search.scss";

export class Search extends Component {
  state = {
    value: "",
    loading: false,
  }

  componentDidUpdate(prevProps, prevState) {
    const {type, pending} = this.props.searchResults;
    const {value, loading} = this.state;
    if (value !== prevState.value || loading) {
      if (pending) {
        !loading && this.setState({loading: true});
        return;
      }
      this.props.loadSearchResults(value, type);
      loading && this.setState({loading: false});
    }
  }

  switchTab = option => {
    this.props.filterByType(option);
    this.props.loadSearchResults(this.state.value, option);
  }

  updateSearchValue = e => {
    this.setState({
      value: e.target.value,
    });
  }

  render() {
    const {value} = this.state;
    const {toggleSearch} = this.props;
    const {isOpen, pending, type, items} = this.props.searchResults;
    return (
      <React.Fragment>
        <TransitionGroup>
          {isOpen &&
            <CSSTransition
              timeout={800}
              classNames="search__input-container"
            >
              <SearchInput
                value={value}
                updateSearchValue={this.updateSearchValue}
                pending={pending}
                close={toggleSearch}
              />
            </CSSTransition>
          }
          {value !== "" && isOpen &&
            <CSSTransition
              timeout={800}
              classNames="search__results"
            >
              <SearchResults
                value={value}
                close={toggleSearch}
                switchTab={this.switchTab}
                type={type}
                items={items}
                pending={pending}
              />
            </CSSTransition>
          }
        </TransitionGroup>
      </React.Fragment>
    );
  }
}

export default connectSearchResults(Search);

Search.propTypes = {
  toggleSearch: PropTypes.func.isRequired,
  filterByType: PropTypes.func.isRequired,
  loadSearchResults: PropTypes.func.isRequired,
  searchResults: PropTypes.object.isRequired,
};