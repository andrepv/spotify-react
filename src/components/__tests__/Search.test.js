import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";

import SearchResults from "components/Search/SearchResults";
import { SearchBtn } from "components/Search/SearchBtn";
import { Search } from "components/Search/Search";
import { initialSearchResultsState } from "reducers/SearchReducer";

describe("<Search />", () => {
  it("renders correctly", () => {
    const props = {
      toggleSearch: jest.fn(),
      filterByType: jest.fn(),
      loadSearchResults: jest.fn(),
      searchResults: {
        ...initialSearchResultsState,
        isOpen: true,
      },
    };
    const component = renderer.create(
      <Search {...props} />
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("should toggle the search bar", () => {
    const props = {
      toggleSearch: jest.fn(),
    };
    const wrapper = shallow(<SearchBtn {...props}/>);
    wrapper.find("div").simulate("click");
    expect(props.toggleSearch.mock.calls.length).toBe(1);

    wrapper.find("div").simulate("click");
    expect(props.toggleSearch.mock.calls.length).toBe(2);
  });

  it("should close search bar", () => {
    const props = {
      toggleSearch: jest.fn(),
      filterByType: jest.fn(),
      loadSearchResults: jest.fn(),
      searchResults: {
        ...initialSearchResultsState,
        isOpen: true,
        pending: true,
      },
    };
    const wrapper = mount(<Search {...props}/>);
    expect(wrapper.find(".search__loader")).toHaveLength(1);

    wrapper.find(".search__close").simulate("click");
    expect(props.toggleSearch.mock.calls.length).toBe(1);
  });

  it("search results bar should open when a user enter a value", () => {
    const props = {
      toggleSearch: jest.fn(),
      filterByType: jest.fn(),
      loadSearchResults: jest.fn(),
      searchResults: {
        ...initialSearchResultsState,
        isOpen: true,
      },
    };
    const wrapper = mount(<Search {...props}/>);
    wrapper.find(".search__input-field").simulate("change", {
      target: { value: "test" },
    });
    expect(wrapper.find(".search__results")).toHaveLength(1);
  });

  it("the click on a tab should toggle search results", () => {
    const props = {
      value: "test",
      type: "track",
      switchTab: jest.fn(),
      items: [{}, {}, {}],
      close: jest.fn(),
      pending: false,
    };
    const wrapper = shallow(<SearchResults {...props}/>);
    expect(wrapper.find(".search__tab_active")).toHaveLength(1);
    expect(wrapper.find(".tracks")).toHaveLength(1);

    wrapper.find(".search__tab").at(3).simulate("click");
    expect(props.switchTab.mock.calls.length).toBe(1);

    wrapper.setProps({
      type: "artist",
      items: [{id: "test", image: "test.png", name:"test"}],
    });
    expect(wrapper.find(".tracks")).toHaveLength(0);
    expect(wrapper.find(".artists")).toHaveLength(1);
  });
});