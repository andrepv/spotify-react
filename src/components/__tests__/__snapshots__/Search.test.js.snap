// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders correctly 1`] = `
<div>
  <div
    className="search__input-container flex-center"
  >
    <div
      className="search__input-wrapper"
    >
      <p>
        What are you looking for?
      </p>
      <div
        className="search__input"
      >
        <input
          className="search__input-field"
          onChange={[Function]}
          placeholder="Start Typing..."
          type="text"
          value=""
        />
        <span
          className="search__close"
          onClick={[MockFunction]}
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
